### ISSUE TITLE
jetbrains.rider 2020.3: can't open projects

### ISSUE THREAD

**Issue by mausch** [NixOS/nixpkgs#106934]
Tried to upgrade jetbrains.rider to 2020.3 (not yet in nixpkgs), getting the following exception when trying to open a .NET solution.
It looks like some .NET code in Rider is trying to guess where libc is? (and obviously can't find it).
Just in case I tried to replace the built-in libleveldb.so with the one in nix but it made no difference.
If nobody here knows I guess I could try asking someone at JetBrains...

```
23:54:03.182 | ERROR | Projected Logger               | IdeaLoggingEvent[message=Could not bind the component argument descriptor on JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesStore. Exception has been thrown by the target of an invocation. The type initializer for 'JetBrains.Application.PersistentMap.LevelDbDriver' threw an exception. The type initializer for 'JetBrains.Application.PersistentMap.LevelDbInterop' threw an exception. The type initializer for 'JetBrains.Util.Interop.NativeDllsLoader' threw an exception. Exception of type 'System.ArgumentOutOfRangeException' was thrown. (Parameter 'linuxLibC') Actual value was Unknown.

--- EXCEPTION #1/7 [ArgumentOutOfRangeException]
ExceptionPath = Root.InnerException.InnerException.InnerException.InnerException.InnerException.InnerException
ClassName = System.ArgumentOutOfRangeException
HResult = COR_E_ARGUMENTOUTOFRANGE=80131502
Source = JetBrains.Platform.Core
ParamName = linuxLibC
ActualValue = Unknown
StackTraceString = “
  at JetBrains.Util.NativeDllEx.ToNativeDir(LinuxLibC linuxLibC)
     at JetBrains.Util.NativeDllEx.GetNativeDir(Postfix postfix, Platform platform, ProcessorArchitecture architecture, Nullable`1 linuxLibC, Boolean debug)
     at JetBrains.Util.NativeDllEx.GetNativeDir(Postfix postfix, ProcessorArchitecture architecture, Boolean debug)
     at JetBrains.Util.Interop.NativeDllsLoader.GetDllDirectory(FileSystemPath root, ProcessorArchitecture architecture, Boolean debug)
     at JetBrains.Util.Interop.NativeDllsLoader.GetDllDirectory(ProcessorArchitecture architecture, Boolean debug)
     at JetBrains.Util.Interop.NativeDllsLoader..cctor()
”

--- Outer ---

--- EXCEPTION #2/7 [TypeInitializationException]
Message = “The type initializer for 'JetBrains.Util.Interop.NativeDllsLoader' threw an exception.”
ExceptionPath = Root.InnerException.InnerException.InnerException.InnerException.InnerException
ClassName = System.TypeInitializationException
InnerException = “Exception #1 at Root.InnerException.InnerException.InnerException.InnerException.InnerException.InnerException”
HResult = COR_E_TYPEINITIALIZATION=80131534
Source = JetBrains.Platform.Util
TypeName = JetBrains.Util.Interop.NativeDllsLoader
StackTraceString = “
  at JetBrains.Util.Interop.NativeDllsLoader.LoadDll(Lifetime lifetime, RelativePath relativePathWithoutSuffix, Options options)
     at JetBrains.Application.PersistentMap.LevelDbInterop.InitIfNecessary()
     at JetBrains.Application.PersistentMap.LevelDbInterop..cctor()
”

--- Outer ---

--- EXCEPTION #3/7 [TypeInitializationException]
Message = “The type initializer for 'JetBrains.Application.PersistentMap.LevelDbInterop' threw an exception.”
ExceptionPath = Root.InnerException.InnerException.InnerException.InnerException
ClassName = System.TypeInitializationException
InnerException = “Exception #2 at Root.InnerException.InnerException.InnerException.InnerException.InnerException”
HResult = COR_E_TYPEINITIALIZATION=80131534
Source = JetBrains.Platform.Shell
TypeName = JetBrains.Application.PersistentMap.LevelDbInterop
StackTraceString = “
  at JetBrains.Application.PersistentMap.LevelDbInterop.InitIfNecessary()
     at JetBrains.Application.PersistentMap.LevelDbDriver..ctor()
     at JetBrains.Application.PersistentMap.LevelDbDriver..cctor()
”

--- Outer ---

--- EXCEPTION #4/7 [TypeInitializationException]
Message = “The type initializer for 'JetBrains.Application.PersistentMap.LevelDbDriver' threw an exception.”
ExceptionPath = Root.InnerException.InnerException.InnerException
ClassName = System.TypeInitializationException
Data.ThreadLocalDebugInfo = “DelayedInitializationStrategy.Dispatch: JetBrains.ProjectModel.FakeSolutionProvider [Singleton, Null]”
InnerException = “Exception #3 at Root.InnerException.InnerException.InnerException.InnerException”
HResult = COR_E_TYPEINITIALIZATION=80131534
Source = JetBrains.UsageStatistics
TypeName = JetBrains.Application.PersistentMap.LevelDbDriver
StackTraceString = “
  at JetBrains.Application.PersistentMap.DbDriverManager.GetDriver()
     at JetBrains.Application.PersistentMap.ShellCaches.CreateDb()
     at JetBrains.Application.PersistentMap.ShellCaches.LazyInit()
     at JetBrains.Application.PersistentMap.ShellCaches.get_Db()
     at JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesLoader.Cache..ctor(Lifetime lifetime, ShellCaches shellCaches)
     at JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesLoader..ctor(Lifetime lifetime, ILogger logger, IShellProcessStartInfoPatcher patcher, ShellCaches shellCaches)
”

--- Outer ---

--- EXCEPTION #5/7 [TargetInvocationException]
Message = “Exception has been thrown by the target of an invocation.”
ExceptionPath = Root.InnerException.InnerException
ClassName = System.Reflection.TargetInvocationException
Data.ThreadLocalDebugInfo = “DelayedInitializationStrategy.Dispatch: JetBrains.ProjectModel.FakeSolutionProvider [Singleton, Null]”
InnerException = “Exception #4 at Root.InnerException.InnerException.InnerException”
HResult = COR_E_TARGETINVOCATION=80131604
Source = System.Private.CoreLib
StackTraceString = “
  at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
     at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
”

--- Outer ---

--- EXCEPTION #6/7 [InvalidOperationException]
Message = “Could not bind the component argument descriptor on JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesStore.”
ExceptionPath = Root.InnerException
ClassName = System.InvalidOperationException
Data.ArgumentIndex = 0
Data.Origin = JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesStore
InnerException = “Exception #5 at Root.InnerException.InnerException”
HResult = COR_E_INVALIDOPERATION=80131509

--- Outer ---

--- EXCEPTION #7/7 [LoggerException]
Message = “Could not bind the component argument descriptor on JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesStore.”
ExceptionPath = Root
ClassName = JetBrains.Util.LoggerException
InnerException = “Exception #6 at Root.InnerException”
HResult = COR_E_APPLICATION=80131600
StackTraceString = “
  at JetBrains.Util.ILoggerEx.LogException(ILogger this, LoggingLevel level, Exception exception, ExceptionOrigin exceptionOrigin, String comment)
     at JetBrains.Util.ILoggerEx.Error(ILogger this, Exception exception, ExceptionOrigin origin, String comment)
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.EnumerableValueResolver.EnumerableComponentDescriptor`1.GetEnumerator()+MoveNext()
     at JetBrains.Application.Components.EnumerableValueResolver.EnumerableComponentDescriptor`1.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.SignatureResolution.BindArguments(List`1 argumentDescriptors, Object origin)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstanceOf(Type type, IValueResolveContext context)
     at JetBrains.Application.Extensibility.PartCatalogTypeComponentDescriptor.CreateInstance(IValueResolveContext context)
     at JetBrains.Application.Components.SingletonDescriptor.CreateInstanceChecked(OnError onError, IComponentContainer container)
     at JetBrains.Application.Components.SingletonDescriptor.GetValue()
     at JetBrains.Application.Components.DelayedInitializationStrategy.<Dispatch>b__8_1()
     at JetBrains.Threading.ReentrancyGuard.Execute(String name, Action action)
     at JetBrains.Threading.ReentrancyGuard.ExecutePendingActions()
     at JetBrains.Threading.JetDispatcher.Closure.Execute()
     at JetBrains.Threading.JetDispatcher.Run()
     at JetBrains.ReSharper.Host.Product.RiderMain.Main(Lifetime lifetime, IRiderSettings settings, ILogger logger, ProtocolComponent protocol, RdShellModel shellModel, ApplicationShutdownRequests shutdownRequests)
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
     at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
     at JetBrains.Application.Environment.RunsPublicStaticIntMain.<>c__DisplayClass0_0.<.ctor>b__0()
     at JetBrains.Application.Environment.RunsPublicStaticIntMain.<>c__DisplayClass0_2.<.ctor>b__5()
     at JetBrains.Util.Logging.Logger.Catch(Action action)
     at JetBrains.Application.Threading.IThreadingEx.<>c__DisplayClass18_1.<ExecuteOrQueueWhenNotGuarded>b__4()
     at JetBrains.Threading.JetDispatcher.Closure.Execute()
     at JetBrains.Threading.JetDispatcher.Run(Func`1 condition, TimeSpan timeout, Boolean bThrowOnTimeout)
     at JetBrains.Application.Environment.IJetHostEx.RunHostMessageLoop(IComponentContainer containerEnv)
     at JetBrains.Application.Environment.HostParameters.MessagePumpMainLoopHostMixin.JetBrains.Application.Environment.HostParameters.IRunMainLoopHostMixin.RunMainLoop(ComponentContainer containerEnv)
     at JetBrains.Application.Environment.HostParameters.JetHostParametersCaller.RunMainLoop(ComponentContainer containerEnv)
     at JetBrains.Application.Environment.JetEnvironment.InternalRun(JetHostParametersCaller host, ComponentContainer containerEnv)
     at JetBrains.Application.Environment.JetEnvironment.CreateAndRun(Full hostparams)
     at JetBrains.Application.Environment.JetHostItems.CreateAndRun(Full input)
     at JetBrains.ReSharper.Host.Product.RiderBackendProgram.MainX(Assembly assembly, HostInfo hostInfo, String[] args)
     at JetBrains.ReSharper.Host.Product.RiderBackendProgram.<>c.<Main>b__3_0(Lifetime lifetime)
     at JetBrains.Lifetimes.Lifetime.Using[T](Func`2 action)
     at JetBrains.ReSharper.Host.Product.RiderBackendProgram.Main()
”
, throwable=com.jetbrains.rdclient.util.BackendException: Could not bind the component argument descriptor on JetBrains.ReSharper.Host.Features.Toolset.Detecting.BuildToolWellKnownPropertiesStore. Exception has been thrown by the target of an invocation. The type initializer for 'JetBrains.Application.PersistentMap.LevelDbDriver' threw an exception. The type initializer for 'JetBrains.Application.PersistentMap.LevelDbInterop' threw an exception. The type initializer for 'JetBrains.Util.Interop.NativeDllsLoader' threw an exception. Exception of type 'System.ArgumentOutOfRangeException' was thrown. (Parameter 'linuxLibC') Actual value was Unknown.
```

**Comment by evenbrenden** [comment]
I'm seeing this too.

**Comment by mausch** [comment]
I just logged this in YouTrack: https://youtrack.jetbrains.com/issue/RIDER-55371

**Comment by mausch** [comment]
Workaround:

```
nix-shell -E 'with import <nixpkgs> {}; (buildFHSUserEnv { 
  name = "rider-fhs"; 
  targetPkgs = p: [
    (p.jetbrains.rider.overrideDerivation (oldAttrs: {
      name = "rider-2020.3";
      src = fetchurl {
        url = https://download-cf.jetbrains.com/rider/JetBrains.Rider-2020.3.tar.gz;
        sha256 = "2483bea98ae598564824692cf0bf08a50a3f682cf617069737d89dcf761c1c20";
      };
    }))
  ]; 
}).env'
```

**Comment by BananchickPasha** [comment]
It seems that this bug is fixed in 2020.3.1. Should we close issue?

**Comment by mausch** [comment]
Yep! A simple upgrade works now https://github.com/NixOS/nixpkgs/pull/107757

BTW I tried upgrading to .NET 5 as JetBrains says that's what they use now but that crashes the backend :man_shrugging:
